flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/DynamodbStreamsProxy.java

```diff
@@ -0,0 +1,222 @@
+		ClientConfiguration awsClientConfig = new ClientConfigurationFactory().getConfig();
+		setAwsClientConfigProperties(awsClientConfig, configProps);
+
+		AWSCredentialsProvider credentials = getCredentialsProvider(configProps);
+		awsClientConfig.setUserAgentPrefix(
+				String.format(
+						USER_AGENT_FORMAT,
+						EnvironmentInformation.getVersion(),
+						EnvironmentInformation.getRevisionInformation().commitId));
+
+		AmazonDynamoDBStreamsAdapterClient adapterClient =
+				new AmazonDynamoDBStreamsAdapterClient(credentials, awsClientConfig);
+
+		if (configProps.containsKey(AWS_ENDPOINT)) {
+			adapterClient.setEndpoint(configProps.getProperty(AWS_ENDPOINT));
+		} else {
+			adapterClient.setRegion(Region.getRegion(
+					Regions.fromName(configProps.getProperty(AWS_REGION))));
+		}
+
+		return adapterClient;
+	}
+
+	@Override
+	public GetShardListResult getShardList(
+			Map<String, String> streamNamesWithLastSeenShardIds) throws InterruptedException {
+		GetShardListResult result = new GetShardListResult();
+
+		for (Map.Entry<String, String> streamNameWithLastSeenShardId :
+				streamNamesWithLastSeenShardIds.entrySet()) {
+			String stream = streamNameWithLastSeenShardId.getKey();
+			String lastSeenShardId = streamNameWithLastSeenShardId.getValue();
+			result.addRetrievedShardsToStream(stream, getShardsOfStream(stream, lastSeenShardId));
+		}
+		return result;
+	}
+
+	private List<StreamShardHandle> getShardsOfStream(
+			String streamName,
+			@Nullable String lastSeenShardId)
+			throws InterruptedException {
+		List<StreamShardHandle> shardsOfStream = new ArrayList<>();
+
+		DescribeStreamResult describeStreamResult;
+		do {
+			describeStreamResult = describeStream(streamName, lastSeenShardId);
+			List<Shard> shards = describeStreamResult.getStreamDescription().getShards();
+			for (Shard shard : shards) {
+				shardsOfStream.add(new StreamShardHandle(streamName, shard));
+			}
+
+			if (shards.size() != 0) {
+				lastSeenShardId = shards.get(shards.size() - 1).getShardId();
+			}
+		} while (describeStreamResult.getStreamDescription().isHasMoreShards());
+
+		return shardsOfStream;
+	}
+
+	/**
+	 * Get metainfo for a Dynamodb stream, which contains information about which shards this
+	 * Dynamodb stream possess.
+	 *
+	 * <p>This method is using a "full jitter" approach described in AWS's article,
+	 * <a href="https://www.awsarchitectureblog.com/2015/03/backoff.html">
+	 *   "Exponential Backoff and Jitter"</a>.
+	 * This is necessary because concurrent calls will be made by all parallel subtask's fetcher.
+	 * This jitter backoff approach will help distribute calls across the fetchers over time.
+	 *
+	 * @param streamName the stream to describe
+	 * @param startShardId which shard to start with for this describe operation
+	 *
+	 * @return the result of the describe stream operation
+	 */
+	private DescribeStreamResult describeStream(String streamName, @Nullable String startShardId)
```

Comment:
Unless there is anything specific to DynamoDB in this implementation,  move it to KinesisProxy. (DescribeStream is standard Kinesis, and it was used for discovery before listShards got  introduced.)

Reply:
OK to move. Just in that case `DescribeStream` will be marked as `protected` at least, in order to be visible to DynamoDBStreamsProxy. 

By keeping it inside the DynamodbStreamsProxy layer this logic is specific for just interacting with DynamodbStreams.

Reply:
Making it protected is fine. Since these are common capabilities of the Kinesis API it is better to keep it in one place.